zebROS_ws/src/behaviors/src/path_to_apriltag.cpp

```diff
@@ -0,0 +1,241 @@
+#include <nav_msgs/Path.h>
+#include <actionlib/client/simple_action_client.h>
+#include <actionlib/client/terminal_state.h>
+#include <path_follower_msgs/PathAction.h>
+#include "base_trajectory_msgs/GenerateSpline.h"
+#include <field_obj/Detection.h>
+#include "tf2_ros/transform_listener.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
+#include "geometry_msgs/PoseWithCovarianceStamped.h"
+#include <optional>
+#include "behavior_actions/PathToAprilTagAction.h"
+#include <sensor_msgs/Imu.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2/LinearMath/Vector3.h>
+#include <actionlib/server/simple_action_server.h>
+
+/*
+"object_id: '1'
+max_objects: 1
+primary_frame_id: 'base_link'
+secondary_object_id: ''
+secondary_max_objects: 0
+secondary_max_distance: 0.0
+secondary_frame_id: ''
+min_radius: 0.0
+endpoint:
+    position: {x: 0.0, y: 0.0, z: 0.0}
+    orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}"
+*/
+
+trajectory_msgs::JointTrajectoryPoint generateTrajectoryPoint(double x, double y, double rotation)
+{
+	trajectory_msgs::JointTrajectoryPoint point;
+	point.positions.resize(3);
+	point.positions[2] = rotation;
+	point.positions[0] = x;
+	point.positions[1] = y;
+	return point;
+}
+
+double getYaw(const geometry_msgs::Quaternion &o) {
+    tf2::Quaternion q;
+    tf2::fromMsg(o, q);
+    tf2::Matrix3x3 m(q);
+    double r, p, y;
+    m.getRPY(r, p, y);
+    return y;
+}
+
+struct Point2D {
+    double x;
+    double y;
+};
+
+double angleBetween(Point2D one, Point2D two) {
+    return atan2(two.y - one.y, two.x - one.x);
+}
+
+class PathToAprilTagAction
+{
+protected:
+
+    ros::NodeHandle nh_;
+    ros::NodeHandle nh_params_;
+    actionlib::SimpleActionServer<behavior_actions::PathToAprilTagAction> as_; // NodeHandle instance must be created before this line. Otherwise strange error occurs.
+    std::string action_name_;
+    // create messages that are used to published feedback/result
+    behavior_actions::PathToAprilTagFeedback feedback_;
+    behavior_actions::PathToAprilTagResult result_;
+    tf2_ros::Buffer tf_buffer_;
+    tf2_ros::TransformListener tf_listener_;
+    double latestImuZ_;
+    double timeout_;
+    field_obj::Detection latest_;
+    actionlib::SimpleActionClient<path_follower_msgs::PathAction> ac_;
+    ros::ServiceClient client_;
+    ros::Subscriber sub_;
+    ros::Subscriber subImu_;
+
+public:
+
+    PathToAprilTagAction(std::string name) :
+        nh_params_(nh_, "path_to_apriltag"),
+        as_(nh_, name, boost::bind(&PathToAprilTagAction::executeCB, this, _1), false),
+        action_name_(name),
+        tf_listener_(tf_buffer_),
+        ac_("/path_follower/path_follower_server", true),
+        client_(nh_.serviceClient<base_trajectory_msgs::GenerateSpline>("/path_follower/base_trajectory/spline_gen")),
+        sub_(nh_.subscribe<field_obj::Detection>("/tf_object_detection/tag_detection_world", 1, boost::bind(&PathToAprilTagAction::callback, this, _1))),
```

Comment:
There's a simpler constructor for subscribers where the callback is a member function : 
`("/frcrobot_jetson/talon_states", 1, &ShooterAction2022::talonStateCallback, this)`